### PR TITLE
[MU4] Plugins actualization

### PR DESCRIPTION
**Pull request by Eism**
Resolves: #10812

### CODE REVIEW

**Comment by jeetee** [comment]
Is the ABC Import(Export) plugin coming in a different PR?
If there won't be a musescore hosted webservice like ABC_import.qml used to use and there is an expectancy to use local tools instead; will MuseScore bundle those tools?

**Comment by Eism** [comment]
@jeetee 
The best solution would be to make requests to the server. Do you know why the server is not available?

**Comment by Jojo-Schmitz** [comment]
Or integrate https://musescore.org/en/project/abcimpex, as #10812 suggests

**Comment by jeetee** [comment]
An advantage to the offline approach is that it works when offline (hello, I'm captain obvious!).
The CON to me is having to provide/support the local version of the tool then (including the eventual python environment) which seems like a fairly big bundling on our side.

The advantage of the server approach is ease of plugin installation for the user. If we go that route I feel hosting the service at .org or .com makes a lot of sense, to be able to fix server issues like the current plugin is having.
Self-hosting might also open up the ability for export as well?

Another option (the most clean and preferred imho, but almost certainly not for 4.0) is to translate that tool into C++ and make ABC a built-in import/export format.

**Comment by Eism** [comment]
@jeetee I agree with you
@Tantacrul Do we need this plugin in 4.0 or can we delay the implementation?

**Comment by Tantacrul** [comment]
> @jeetee I agree with you
> 
> @Tantacrul Do we need this plugin in 4.0 or can we delay the implementation?

I think it is fine to delay this one until you have an ideal technical solution. Can you create a separate task for it and we'll bump to 4.x?

**Comment by MarcSabatella** [comment]
I have only minimal experience with AppSpot, but I suspect it’s something very simple to recreate.  I know something similar can be done through GitHub - GitHub.io, I think?  I can try to look into this.

File: src/plugins/api/qmlpluginapi.h [hunk 1]
```diff
@@ -81,19 +81,21 @@ class PluginAPI : public mu::engraving::QmlPlugin
     INJECT(plugins, mu::context::IGlobalContext, context)
 
     /** Path where the plugin is placed in menu */
-    Q_PROPERTY(QString menuPath READ menuPath WRITE setMenuPath)
```
Comment (RomanPudashkin) [inline src/plugins/api/qmlpluginapi.h:84]: I would recommend not changing the old API. It definitely requires some redesing, but that's not for this PR

File: src/plugins/view/pluginview.cpp [hunk 2]
```diff
@@ -103,6 +136,9 @@ void PluginView::run()
     //connect(m_view, &QQuickView::closing, this, &PluginView::finished);
     connect(m_view, SIGNAL(closing(QQuickCloseEvent*)), this, SIGNAL(finished()));
 
+    if (pluginHasUi()) {
```
Comment (RomanPudashkin) [inline src/plugins/view/pluginview.cpp:139]: I would recommend rewriting this code as follows:

```
void PluginView::run()
{
    if (!pluginHasUi()) {
        m_qmlPlugin->runPlugin();
        return;
    }
    
    destroyView();
    m_view = new QQuickView(engine(), nullptr);
    ...
}
```

File: src/plugins/view/pluginview.cpp [hunk 3]
```diff
@@ -35,6 +35,12 @@ PluginView::PluginView(const QUrl& url, QObject* parent)
 {
     m_component = new QQmlComponent(engine(), url);
     m_qmlPlugin = qobject_cast<mu::engraving::QmlPlugin*>(m_component->create());
+
+    connect(m_qmlPlugin, &mu::engraving::QmlPlugin::close, [this]() {
```
Comment (RomanPudashkin) [inline src/plugins/view/pluginview.cpp:39]: I would suggest renaming this signal to closeRequested

**Review by RomanPudashkin — APPROVED** [review]

**Comment by jeetee** [comment]
FYI: I just got a basic python "hello world" enabled on my server, so I might be able to run the convertor (perhaps even both directions) in the coming week(s) via https://musescore.jeetee.net/

I'd still prefer a .org hosted version, but due to the amount of users complaining about the current plugin failures it can act as a duct tape stop-gap intermediate solution

**Comment by Tantacrul** [comment]
@jeetee - I think we can sort that out!

DM me and we'll figure out the details

**Pull request merged**